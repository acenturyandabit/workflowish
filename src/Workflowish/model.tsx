import * as localforage from "localforage";
import * as React from "react";
import { v4 as uuidv4 } from 'uuid';
export type ItemTreeNode = {
    id: string,
    data: string,
    children: ItemTreeNode[],
    collapsed: boolean
}

type FlatItemData = {
    data: string,
    parentId: string,
    indexInParent: number,
    collapsed: boolean
}

type FlatItemBlob = Record<string, FlatItemData>;
export const makeNewUniqueKey = (): string => {
    return (Date.now().toString()) + uuidv4();
}

const defaultEmptyArray = () => [{
    id: makeNewUniqueKey(),
    data: "",
    children: [],
    collapsed: false
}];

export const useSavedItems = (): [Array<ItemTreeNode>, React.Dispatch<React.SetStateAction<Array<ItemTreeNode>>>] => {
    const [todoItems, setTodoItems] = React.useState<Array<ItemTreeNode>>(defaultEmptyArray())
    React.useEffect(() => {
        (async () => {
            const localForageTodoItems: FlatItemBlob | null = await localforage.getItem<FlatItemBlob>("items")
            if (localForageTodoItems) {
                setTodoItems(buildTree(localForageTodoItems));
            }
        })()
    }, [])

    React.useEffect(() => {
        localforage.setItem<FlatItemBlob>("items", fromTree(todoItems));
    }, [todoItems])

    return [todoItems, setTodoItems];
}

const buildTree = (flatItemBlob: FlatItemBlob): Array<ItemTreeNode> => {
    const treeRootsArray: Array<ItemTreeNode> = [];
    const treeConstructorRecord: Record<string,
        { indexInParent: number, children: Array<ItemTreeNode & { indexInParent: number }> } & ItemTreeNode
    > = {};
    // First pass: instantiation. 
    for (const nodeId in flatItemBlob) {
        treeConstructorRecord[nodeId] = {
            id: nodeId,
            data: flatItemBlob[nodeId].data,
            indexInParent: flatItemBlob[nodeId].indexInParent,
            children: [],
            collapsed: flatItemBlob[nodeId].collapsed
        }
    }
    // Second pass: Parent assignment
    for (const nodeId in flatItemBlob) {
        const parentId = flatItemBlob[nodeId].parentId;
        const currentNode = treeConstructorRecord[nodeId];
        if (parentId in treeConstructorRecord) {
            treeConstructorRecord[parentId].children.push(currentNode);
        } else {
            // Lost OR root items
            treeRootsArray.push(currentNode);
        }
    }
    // Clean up the ordering
    for (const nodeId in flatItemBlob) {
        const currentNode = treeConstructorRecord[nodeId];
        currentNode.children.sort((a, b) => a.indexInParent - b.indexInParent);
    }
    // Remove indexInParent
    treeRootsArray.forEach((n: ItemTreeNode & { indexInParent?: number }) => {
        delete n.indexInParent;
    })
    if (treeRootsArray.length == 0) return defaultEmptyArray();
    else return treeRootsArray;
}

const fromTree = (itemTreeArray: Array<ItemTreeNode>): FlatItemBlob => {
    const flatBlob: FlatItemBlob = {};
    const unrollItems = (parentId: string, nodes: ItemTreeNode[]) => {
        nodes.forEach((n, idx) => {
            flatBlob[n.id] = {
                data: n.data,
                parentId: parentId,
                indexInParent: idx,
                collapsed: n.collapsed
            }
            unrollItems(n.id, n.children);
        })
    }
    unrollItems("", itemTreeArray);
    return flatBlob;
}