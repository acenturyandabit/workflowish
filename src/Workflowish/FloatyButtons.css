.floatyButtons{
    position: absolute;
    display: flex;
    left: 0px;
    right: 0px;
    bottom: 0px;
    transform-origin: left bottom;
    transform: translate(0px, 0px) scale(1);
}

.floatyButtons button{
    flex: 1 1 auto;
    height: 40px;
}

.floatyButtons button.pressed{
    background:lightblue;
}
.floatyButtons button.heavyPressed{
    color:white;
    background:darkblue;
}