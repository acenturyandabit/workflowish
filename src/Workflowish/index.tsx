import * as React from "react";
import { BaseStoreDataType } from "~CoreDataLake";
import { makeListActions, TreeNodeArrayGetSetter, TreeNodesGetSetter } from "./mvc/controller";
import { FocusedActionReceiver, dummyFocusedActionReciever } from "./mvc/focusedActionReceiver"
import Item, { FocusActions } from "./Item"
import { ItemTreeNode, getTransformedDataAndSetter, TransformedDataAndSetter, virtualRootId, TransformedData } from "./mvc/model"
import { isMobile } from '~util/isMobile';
import { FloatyButtons } from "./Subcomponents/FloatyButtons";
import OmnibarWrapper from "./Subcomponents/OmnibarWrapper";
import ContextMenu from "./Subcomponents/ContextMenu";
import { ModelContext, RenderTimeContext } from "./mvc/context";

export default (props: {
    data: BaseStoreDataType,
    updateData: React.Dispatch<React.SetStateAction<BaseStoreDataType>>
}) => {
    const transformedDataAndSetter = getTransformedDataAndSetter({ data: props.data, updateData: props.updateData });
    const [focusedActionReceiver, setFocusedActionReceiver] = React.useState<FocusedActionReceiver>(dummyFocusedActionReciever);
    const itemsRefDictionary = React.useRef<Record<string, FocusActions>>({});
    const [lastFocusedItem, setLastFocusedItem] = React.useState<string>("");

    const [showIds, setShowIds] = React.useState<boolean>(false);
    React.useEffect(AltShouldToggleShowIds(setShowIds), []);

    return <div style={{ height: "100%", display: "flex", flexDirection: "column" }}>
        <ContextMenu></ContextMenu>
        <ModelContext.Provider value={transformedDataAndSetter.transformedData.rootNode}>
            <div style={{ margin: "0 5px 10px 5px", flex: "1 0 auto" }}>
                <OmnibarWrapper
                    itemRefsDictionary={itemsRefDictionary.current}
                    transformedDataAndSetter={transformedDataAndSetter}
                    lastFocusedItem={lastFocusedItem}
                >
                    <ItemsList
                        showIds={showIds}
                        itemRefsDictionary={itemsRefDictionary.current}
                        setFocusedActionReceiver={setFocusedActionReceiver}
                        lastFocusedItem={lastFocusedItem}
                        setLastFocusedItem={setLastFocusedItem}
                        transformedDataAndSetter={transformedDataAndSetter}
                    ></ItemsList>
                </OmnibarWrapper>
            </div>
        </ModelContext.Provider>
        {isMobile() ? <FloatyButtons focusedActionReceiver={focusedActionReceiver}></FloatyButtons> : null}
    </div>
};

const AltShouldToggleShowIds = (setShowIds: React.Dispatch<React.SetStateAction<boolean>>) => {
    const altModifyToggle = (evt: KeyboardEvent) => {
        setShowIds(evt.altKey && evt.shiftKey);
        if (evt.key == "Alt") evt.preventDefault();
    }
    window.addEventListener("keydown", altModifyToggle);
    window.addEventListener("keyup", altModifyToggle);
    return () => {
        window.removeEventListener("keydown", altModifyToggle);
        window.removeEventListener("keyup", altModifyToggle);
    }
}

const ItemsList = (
    props: {
        setFocusedActionReceiver: React.Dispatch<React.SetStateAction<FocusedActionReceiver>>,
        itemRefsDictionary: Record<string, FocusActions>,
        transformedDataAndSetter: TransformedDataAndSetter,
        lastFocusedItem: string,
        setLastFocusedItem: React.Dispatch<React.SetStateAction<string>>
        showIds: boolean
    }
) => {
    const itemTree = React.useContext<ItemTreeNode>(ModelContext);
    const nullSizedArrayForRefs = Array(itemTree.children.length).fill(null);
    const topLevelNullFunction = () => {
        // Top level cannot take focus
    }

    const setFocusedItem = (focusedActionReceiver: FocusedActionReceiver, focusItemKey: string) => {
        props.setFocusedActionReceiver(focusedActionReceiver);
        props.setLastFocusedItem(focusItemKey);
    }

    const itemsRefArray = React.useRef<Array<FocusActions | null>>(nullSizedArrayForRefs);
    return <RenderTimeContext.Provider value={{
        currentFocusedItem: props.lastFocusedItem
    }}>{itemTree.children.map((item, ii) => {
        return (<Item
            key={ii}
            styleParams={{
                showId: props.showIds,
                emptyList: itemTree.children.length == 1 && item.data == ""
            }}
            item={item}
            pushRef={(ref: FocusActions) => itemsRefArray.current[ii] = ref}
            pushRefGlobal={(ref: FocusActions, id: string) => { props.itemRefsDictionary[id] = ref }}
            setThisAsFocused={setFocusedItem}
            actions={makeListActions({
                siblingsFocusActions: itemsRefArray,
                currentSiblingIdx: ii,
                getSetSiblingArray: (t: TreeNodeArrayGetSetter) => {
                    props.transformedDataAndSetter.setItemsByKey((oldData: TransformedData) => {
                        const newChildren = t(oldData.rootNode.children);
                        // There was an implicit constraint that the _ENTIRE TREE_ shall also be updated
                        // in the old fromTree-based getSetSiblingArray. When we delete getSetSiblingArray
                        // we should also delete this.
                        const nodeStack: ItemTreeNode[] = newChildren;
                        // initialize flatTree with the top level children + the virtualRoot
                        const flatTree: Record<string, ItemTreeNode> = newChildren.reduce((aggregate, itm) =>
                            ({ ...aggregate, [itm.id]: itm }), {} as Record<string, ItemTreeNode>);
                        flatTree[virtualRootId] = { ...oldData.rootNode, children: [...newChildren] }
                        while (nodeStack.length) {
                            const top = nodeStack.shift()
                            if (top) {
                                const noDuplicateChildren = top.children.filter(child => {
                                    if (child.id in flatTree) {
                                        // Log an error and ignore the duplicate
                                        console.error(`Duplicate node ${top.id}! Removing from this parent and moving on.`);
                                        return false;
                                    } else {
                                        flatTree[child.id] = child;
                                        nodeStack.push({ ...child, markedForCleanup: top.markedForCleanup || child.markedForCleanup })
                                        return true;
                                    }
                                });
                                if (noDuplicateChildren.length != top.children.length) {
                                    top.children = noDuplicateChildren;
                                    top.lastModifiedUnixMillis = Date.now();
                                }
                            }
                        }
                        return flatTree;
                    })
                },
                unindentCaller: () => {
                    // cannot unindent at root level
                },
                focusItem: (id: string) => {
                    props.itemRefsDictionary[id].focusThis();
                },
                parentFocusActions: {
                    triggerFocusFromAbove: topLevelNullFunction,
                    triggerFocusFromBelow: topLevelNullFunction,
                    focusThis: () => itemsRefArray.current[ii]?.focusThis(),
                    focusThisEnd: topLevelNullFunction,
                    focusRecentlyIndentedItem: topLevelNullFunction,
                    focusMyNextSibling: () => {
                        if (ii < itemTree.children.length - 1) {
                            itemsRefArray.current[ii + 1]?.focusThis()
                        } else {
                            itemsRefArray.current[ii]?.focusThis()
                        }
                    },
                    scrollThisIntoView: topLevelNullFunction
                },
                disableDelete: () => (itemTree.children.length == 1),
                getSetItems: (keys: string[], getSetter: TreeNodesGetSetter) => {
                    props.transformedDataAndSetter.setItemsByKey((oldData: TransformedData) => {
                        // TODO: Remove all instances of this getSetItems interface and replace it with just setItemsByKey
                        const oldItems = keys.map(key => oldData.keyedNodes[key])
                        const newNodes = getSetter(oldItems);
                        return newNodes.reduce((nodeDict, node) => {
                            nodeDict[node.id] = node;
                            return nodeDict;
                        }, {} as Record<string, ItemTreeNode>);
                    });
                },
                thisItem: itemTree
            })}
        ></Item >)
    })}</RenderTimeContext.Provider>
}