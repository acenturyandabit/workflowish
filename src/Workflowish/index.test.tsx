/**
 * @jest-environment jsdom
 */

// https://github.com/nareshbhatia/react-testing-techniques/blob/main/docs/fireEvent-vs-userEvent.md
import * as renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import * as React from "react";
import { it, expect, jest } from '@jest/globals';
import Workflowish from "./index";
import { BaseStoreDataType, jestSetMakeUniqueKey } from '~CoreDataLake';
import { resolveAllDocuments } from '~CoreDataLake';
import { fromNestedRecord } from './mvc/firstTimeDoc';
import { FlatItemBlob, FlatItemData, fromTree, virtualRootId } from './mvc/model';

jestSetMakeUniqueKey(() => "newItem");

export const makeMockData = (mockData: BaseStoreDataType): [
    (newData:
        BaseStoreDataType |
        ((oldData: BaseStoreDataType) => BaseStoreDataType)
    ) => void,
    () => BaseStoreDataType
] => {
    let dataSetByConsumer = mockData;
    const mockUpdateData = (newData:
        BaseStoreDataType |
        ((oldData: BaseStoreDataType) => BaseStoreDataType)
    ) => {
        let newDataToSet: BaseStoreDataType;
        if (typeof newData == "function") {
            newDataToSet = newData(mockData);
        } else {
            newDataToSet = newData;
        }
        dataSetByConsumer = resolveAllDocuments([dataSetByConsumer, newDataToSet]);
    }
    const getDataSetByConsumer = () => dataSetByConsumer;
    return [mockUpdateData, getDataSetByConsumer]
}

it('Creates the new document if no document is provided', () => {
    jest.useFakeTimers();
    jest.spyOn(global, 'setTimeout');

    const initialEmptyData = {};
    const [mockUpdateData, getDataSetByConsumer] = makeMockData(initialEmptyData);
    const component = renderer.create(
        <Workflowish
            data={initialEmptyData}
            updateData={mockUpdateData}
        ></Workflowish>
    )
    const firstRender = component.toJSON();
    expect(firstRender).toMatchSnapshot();

    expect(setTimeout).toHaveBeenCalledTimes(1);
    jest.runAllTimers();

    const secondRender = renderer.create(
        <Workflowish
            data={getDataSetByConsumer()}
            updateData={mockUpdateData}
        ></Workflowish>
    )
    const secondRenderJ = secondRender.toJSON();
    expect(secondRenderJ).toMatchSnapshot();
    jest.restoreAllMocks();
    jest.useRealTimers();
})

it('Indents an item correctly when tabbed into a symlink', async () => {

    const user = userEvent.setup({ delay: null }) // https://github.com/testing-library/user-event/issues/833

    const initialData: FlatItemBlob = fromTree(fromNestedRecord({
        data: virtualRootId,
        id: virtualRootId,
        children: [
            {
                data: "symlink target",
                id: "symlink_target"
            },
            {
                data: "intermediate parent",
                id: "intermediate_parent",
                children: [
                    {
                        data: "[LN: symlink_target]",
                        id: "link_item"
                    },
                    {
                        data: "a child",
                        id: "child"
                    }
                ]
            }
        ]
    }));
    const [mockUpdateData, getDataSetByConsumer] = makeMockData(initialData);
    render(
        <Workflowish
            data={initialData}
            updateData={mockUpdateData}
        ></Workflowish>
    );

    await user.click(screen.getByTestId("child"));
    await user.keyboard("{Tab}");
    const symlinkFlat: FlatItemData = getDataSetByConsumer()["symlink_target"] as FlatItemData;
    expect(symlinkFlat.children[0]).toBe("child")

})

it('Unindents an item correctly when tabbed out from a symlink', async () => {

    const user = userEvent.setup({ delay: null }) // https://github.com/testing-library/user-event/issues/833

    const initialData: FlatItemBlob = fromTree(fromNestedRecord({
        data: virtualRootId,
        id: virtualRootId,
        children: [
            {
                data: "symlink target",
                id: "symlink_target",
                children: [{
                    data: "a child",
                    id: "child"
                }]
            },
            {
                data: "intermediate parent",
                id: "intermediate_parent",
                children: [
                    {
                        data: "[LN: symlink_target]",
                        id: "link_item"
                    },
                ]
            }
        ]
    }));
    const [mockUpdateData, getDataSetByConsumer] = makeMockData(initialData);

    render(
        <Workflowish
            data={initialData}
            updateData={mockUpdateData}
        ></Workflowish>
    );

    await user.click(screen.getByTestId("link_item@child"));
    await user.keyboard("{Shift>}{Tab}{/Shift}");
    const newData: FlatItemBlob = getDataSetByConsumer() as FlatItemBlob;
    expect(newData["intermediate_parent"].children[1]).toBe("child")
    expect(newData["symlink_target"].children.length).toBe(0)

})


it("Unindent at root level doesn't explode", async () => {
    const user = userEvent.setup({ delay: null }) // https://github.com/testing-library/user-event/issues/833

    const initialData: FlatItemBlob = fromTree(fromNestedRecord({
        data: virtualRootId,
        id: virtualRootId,
        children: [
            {
                data: "root level item",
                id: "root_level_item"
            }
        ]
    }));
    const expected = JSON.stringify(initialData);
    const [mockUpdateData, getDataSetByConsumer] = makeMockData(initialData);

    render(
        <Workflowish
            data={initialData}
            updateData={mockUpdateData}
        ></Workflowish>
    );

    await user.click(screen.getByTestId("root_level_item"));
    await user.keyboard("{Shift>}{Tab}{/Shift}");
    expect(JSON.stringify(getDataSetByConsumer())).toBe(expected)

})

