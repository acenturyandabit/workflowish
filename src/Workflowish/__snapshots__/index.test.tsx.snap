// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the new document if no document is provided 1`] = `
<div
  style={
    {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
    }
  }
>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="üîç Search"
      style={
        {
          "flex": "1 0 auto",
          "padding": "2px",
        }
      }
      value=""
    />
  </div>
  <div
    style={
      {
        "flex": "1 0 auto",
        "margin": "0 5px 10px 5px",
      }
    }
  >
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0.2em",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚óè
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Welcome to Workflowish!",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      />
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñ∂
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Workflowish is a keyboard-focused nested listing app.",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñº
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "This means you can nest bullet points...",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      >
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0px",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚ñº
              </span>
            </span>
            <div
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Like this,",
                }
              }
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          >
            <span
              className="itemWrapperClass"
            >
              <span
                onContextMenu={[Function]}
                style={
                  {
                    "background": "",
                  }
                }
              >
                <span
                  style={
                    {
                      "color": "white",
                      "cursor": "pointer",
                      "paddingLeft": "0px",
                      "userSelect": "none",
                    }
                  }
                >
                  <span
                    onClick={[Function]}
                  >
                    ‚ñº
                  </span>
                </span>
                <div
                  contentEditable={true}
                  dangerouslySetInnerHTML={
                    {
                      "__html": "And this,",
                    }
                  }
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onInput={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    {
                      "flex": "1 1 auto",
                      "scrollMarginTop": "100px",
                    }
                  }
                />
              </span>
              <div
                style={
                  {
                    "borderLeft": "1px solid white",
                    "marginLeft": "0.5em",
                    "paddingLeft": "5px",
                  }
                }
              >
                <span
                  className="itemWrapperClass"
                >
                  <span
                    onContextMenu={[Function]}
                    style={
                      {
                        "background": "",
                      }
                    }
                  >
                    <span
                      style={
                        {
                          "color": "white",
                          "cursor": "pointer",
                          "paddingLeft": "0.2em",
                          "userSelect": "none",
                        }
                      }
                    >
                      <span
                        onClick={[Function]}
                      >
                        ‚óè
                      </span>
                    </span>
                    <div
                      contentEditable={true}
                      dangerouslySetInnerHTML={
                        {
                          "__html": "And so on.",
                        }
                      }
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onInput={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      style={
                        {
                          "flex": "1 1 auto",
                          "scrollMarginTop": "100px",
                        }
                      }
                    />
                  </span>
                  <div
                    style={
                      {
                        "borderLeft": "1px solid white",
                        "marginLeft": "0.5em",
                        "paddingLeft": "5px",
                      }
                    }
                  />
                </span>
              </div>
            </span>
          </div>
        </span>
      </div>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñº
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "You can edit any item in the list by clicking it and typing :)",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      >
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "You can also add new items by pressing Enter on an existing item; or,",
                }
              }
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Backspace all the way to the end, and then once more, to delete an item.",
                }
              }
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Press CTRL+S to save your changes to your browser. (Other save options are available, read on!)",
                }
              }
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
      </div>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñ∂
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Since lists get complicated, you can also collapse/show lists by focusing on an item and pressing CTRL+UP/DOWN.",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñº
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "Workflowish is inspired by Workflowy (https://workflowy.com/), but is free and open source, with in-browser storage or self-hosting.",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
      <div
        style={
          {
            "borderLeft": "1px solid white",
            "marginLeft": "0.5em",
            "paddingLeft": "5px",
          }
        }
      >
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0px",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚ñ∂
              </span>
            </span>
            <div
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "Workflowish also works on a mobile phone, again with a keyboard-first experience.",
                }
              }
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
        </span>
        <span
          className="itemWrapperClass"
        >
          <span
            onContextMenu={[Function]}
            style={
              {
                "background": "",
              }
            }
          >
            <span
              style={
                {
                  "color": "white",
                  "cursor": "pointer",
                  "paddingLeft": "0.2em",
                  "userSelect": "none",
                }
              }
            >
              <span
                onClick={[Function]}
              >
                ‚óè
              </span>
            </span>
            <div
              contentEditable={true}
              dangerouslySetInnerHTML={
                {
                  "__html": "For self-hosted storage, clone the github repository (https://github.com/acenturyandabit/workflowish) and follow the README.",
                }
              }
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                {
                  "flex": "1 1 auto",
                  "scrollMarginTop": "100px",
                }
              }
            />
          </span>
          <div
            style={
              {
                "borderLeft": "1px solid white",
                "marginLeft": "0.5em",
                "paddingLeft": "5px",
              }
            }
          />
        </span>
      </div>
    </span>
    <span
      className="itemWrapperClass"
    >
      <span
        onContextMenu={[Function]}
        style={
          {
            "background": "",
          }
        }
      >
        <span
          style={
            {
              "color": "white",
              "cursor": "pointer",
              "paddingLeft": "0px",
              "userSelect": "none",
            }
          }
        >
          <span
            onClick={[Function]}
          >
            ‚ñ∂
          </span>
        </span>
        <div
          contentEditable={true}
          dangerouslySetInnerHTML={
            {
              "__html": "As an extra treat for superusers out there, Workflowish also comes with:",
            }
          }
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            {
              "flex": "1 1 auto",
              "scrollMarginTop": "100px",
            }
          }
        />
      </span>
    </span>
  </div>
</div>
`;
